### PR TITLE
SP-6356 / Step Status sprint review fixes

### PR DESCRIPTION
**Pull request by kostyay**
* StepError -> StepStatus
* Code -> codes.Code
* Status -> status.Status
* step.Run -> does os.Exit() with exit code

### CODE REVIEW

**Review by vic3lord — APPROVED** [review]

File: step/status/status.go [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+	Message string `json:"message"`
+	// Verbose error message
+	Verbose string `json:"verbose"`
+}
+
+// OK - successful Status
+var OK = &Status{Code: codes.OK}
+
+// Newf initialize Status with code, message and verbose
+func Newf(code codes.Code, message, verbose string, a ...interface{}) *Status {
```
Comment (alongat) [inline step/status/status.go:22]: Newf? 🤯 didn't know this convention 
like.

File: step/runner.go [hunk 2]
```diff
@@ -0,0 +1,17 @@
+package step
+
+import (
+	"os"
+
+	"github.com/stackpulse/steps-sdk-go/exec"
+)
+
+// Run runs a step and sets the exit code to either `exec.ExitCodeFailure` on error or to `exec.ExitCodeOK` on success
+func Run(step Step) {
```
Comment (hagaishapira) [inline step/runner.go:10]: 👍

File: step/step.go [hunk 3]
```diff
@@ -3,59 +3,65 @@ package step
...
 }
 
-func printOutput(stepOutput map[string]interface{}, stepError *Error) error {
-	if stepError == nil {
-		stepError = ErrSuccess
+func formatInternalError(err error) string {
+	return fmt.Sprintf(`{"%s":{"code":%d,"verbose":"%s"}}`, StatusKey, codes.Internal, err.Error())
+}
+
+func writeOutput(w io.Writer, stepOutput Output, stepStatus *Status) error {
```
Comment (hagaishapira) [inline step/step.go:35]: It doesn't look like this returns an error if `status.Failed(stepStatus)` which is what should happen from the step writer perspective

**Review by hagaishapira — COMMENTED** [review]
Looks really good, check out my one comment and let me know if there is a real issue there.

File: step/status/status_test.go [hunk 4]
```diff
@@ -0,0 +1,29 @@
...
+
+import (
+	"fmt"
+	"testing"
+
+	"github.com/stackpulse/steps-sdk-go/step/codes"
+	"github.com/stretchr/testify/assert"
+)
+
+func TestNewf(t *testing.T) {
```
Comment (talbe) [inline step/status/status_test.go:11]: why not table tests?

File: step/step_test.go [hunk 5]
```diff
@@ -1,164 +1,96 @@
...
-	"log"
-	"os"
+	"fmt"
 	"strings"
 	"testing"
 
+	"github.com/golang/mock/gomock"
+	"github.com/stackpulse/steps-sdk-go/step/codes"
+	"github.com/stackpulse/steps-sdk-go/step/status"
 	"github.com/stretchr/testify/require"
```
Comment (alongat) [inline step/step_test.go:14]: \n

File: testutil/container/matcher/matchers_test.go [hunk 6]
```diff
@@ -317,3 +321,80 @@ func TestStringRegexMatch(t *testing.T) {
...
+
+func TestStepStatus(t *testing.T) {
+
+	tests := []struct {
+		name        string
+		input       *status.Status
+		wantCode    codes.Code
+		wantMessage string
+		wantVerbose string
+		wantError   bool
```
Comment (talbe) [inline testutil/container/matcher/matchers_test.go:333]: it's better to use actual error instead of boolean

File: step/runner_test.go [hunk 7]
```diff
@@ -0,0 +1 @@
+package step
```
Comment (talbe) [inline step/runner_test.go:1]: implement tests here or delete empty file

File: step/status/status_test.go [hunk 8]
```diff
@@ -0,0 +1,29 @@
...
+
+func TestFromError(t *testing.T) {
+	err := fmt.Errorf("error")
+	s := FromError(codes.Internal, "msg", err)
+	assert.Equal(t, s.Code, codes.Internal)
+	assert.Equal(t, s.Message, "msg")
+	assert.Equal(t, s.Verbose, "error")
+}
+
+func TestFailed(t *testing.T) {
```
Comment (talbe) [inline step/status/status_test.go:26]: &Status{} would return Failed true
does it the expected result?

**Review by oriser — COMMENTED** [review]
See comments

File: step/step_test.go [hunk 9]
```diff
@@ -1,164 +1,96 @@
...
 
-	"github.com/Jeffail/gabs/v2"
-
 	"github.com/stackpulse/steps-sdk-go/env"
 	"github.com/stretchr/testify/assert"
 )
 
-type successStep struct {
```
Comment (talbe) [inline step/step_test.go:18]: I wouldn't leave those tests
maybe use this in the runner package

File: step/step.go [hunk 10]
```diff
@@ -3,59 +3,65 @@ package step
...
-	Init() *Error
-	Run() (Output, *Error)
+	Init() *Status
+	Run() (Output, *Status)
 }
 
-func printOutput(stepOutput map[string]interface{}, stepError *Error) error {
-	if stepError == nil {
-		stepError = ErrSuccess
+func formatInternalError(err error) string {
```
Comment (talbe) [inline step/step.go:31]: Consider make it public
orchestrator would use that as fallback in case the step didn't generate the step status

Comment (talbe) [inline step/step.go:31]: nil error would stop the program
check error not nil

File: step/step_test.go [hunk 11]
```diff
@@ -1,164 +1,96 @@
...
+			require.True(t, status.Equals(result.StepStatus, tc.wantStatus))
+			if !status.Failed(tc.stepStatus) {
+				require.Equal(t, result.Output, tc.wantOutput)
+			}
 		})
 	}
 }
+
+func Test_formatInternalError(t *testing.T) {
+	err := fmt.Errorf("error")
```
Comment (talbe) [inline step/step_test.go:88]: Consider table tests this function since you want to make sure that you behave as expected in nil error scenario

**Pull request merged**